src/devices/machine/bl_handhelds_menucontrol.h

```diff
@@ -0,0 +1,64 @@
+// license:BSD-3-Clause
+// copyright-holders:David Haywood
+
+#ifndef MAME_MACHINE_BL_HANDHELDS_MENUCONTROL_H
+#define MAME_MACHINE_BL_HANDHELDS_MENUCONTROL_H
+
+#pragma once
+
+DECLARE_DEVICE_TYPE(BL_HANDHELDS_MENUCONTROL, bl_handhelds_menucontrol_device)
+
+class bl_handhelds_menucontrol_device : public device_t
+{
+public:
+	// construction/destruction
+	bl_handhelds_menucontrol_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	// the chip is the same between systems, but there's some logic not fully understood that is causing off-by-1 errors on some calcs
+	void set_is_unsp_type_hack() { m_is_unsp_type_hack = true; }
+
+	DECLARE_READ_LINE_MEMBER(status_r);
+	DECLARE_READ_LINE_MEMBER(data_r);
+	DECLARE_WRITE_LINE_MEMBER(clock_w);
+	DECLARE_WRITE_LINE_MEMBER(data_w);
+	DECLARE_WRITE_LINE_MEMBER(reset_w);
+
+protected:
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+private:
+	// config
+	bool m_is_unsp_type_hack;
+
+	// internal state
+	int m_menupos;
+	
+	// command handling
+	int m_clockstate;
+	int m_datashifterpos;
+
+	int m_responsebit;
+	uint8_t m_response;
+
+	int m_commandbit;
+	uint8_t m_command;
+
+	void handle_command();
+
+	enum menustate : const int
```

Comment:
It makes no sense to use `const int` as a storage class.

Reply:
I don't see anything in the highlighted block that involves the keyword "const".

Reply:
Last line (line 49): `enum menustate : const int`

It makes no sense to put const/volatile qualifiers in an enum storage class.

Reply:
interestingly the coding standards document for a recent work project explicitly requires me to do that, on the grounds that anything in an enum is basically a constant.

Reply:
The storage class is used to force a particular type for the enum.  For example: `enum bar : uint8_t` means force the enum to be represented as an 8-bit unsigned value.  Putting `const` or `volatile` there has no practical effect, only the size/signedness/alignment requirements are used.

The enumerated value names are always constants, and need to be initialised with `constexpr` expressions.  They have no storage locations, you can’t take the address of them, and hence they can’t be used in lvalue context and ODR doesn’t apply to them.

The default if no storage class is specified is to use the smallest of `int`, `long` or `long long` that can represent all the enumerated values.  This behaviour is mostly backwards-compatible with older versions of C++.